src/main/java/io/vertx/core/streams/impl/OutboundWriteQueue.java

```diff
@@ -0,0 +1,361 @@
+ * When a method returns a {@link #DRAIN_REQUIRED_MASK} signal, the {@link #drain()} should be called by the consumer thread
+ *
+ * <ul>
+ *   <li>A consumer method ({@link #add}, {@link #drain}) emits this signal when the {@link #consumer} refuses an element,
+ *   therefore {@link #drain()} should be called when the consumer can accept element again</li>
+ *   <li>A producer method ({@link #submit}) emits this signal when it acquired the ownership of the queue to {@link #drain} the
+ *   queue.</li>
+ * </ul>
+ *
+ * <h3>Back-pressure</h3>
+ *
+ * <p>Producers emission flow cannot reliably be controlled by the consumer back-pressure probe since producers
+ * emissions are transferred to the consumer thread: the consumer back-pressure controller does not take in account the
+ * inflight elements between producers and consumer. This queue is designed to provide reliable signals to control
+ * producers emission based on the consumer back-pressure probe and the number of inflight elements between
+ * producers and the consumer.</p>
+ *
+ * The queue maintains an internal queue of elements initially empty and can be filled when
+ * <ul>
+ *   <li>producer threads {@link #submit)} to the queue above the {@link #highWaterMark}</li>
+ *   <li>the {@link #consumer} refuses an element</li>
+ * </ul>
+ *
+ * <p>When the internal queue grows above the {@link #highWaterMark}, the queue is considered as {@code unwritable}.
+ * {@link #add}/{@link #submit} methods return {@link #QUEUE_UNWRITABLE_MASK} to signal producers should stop emitting.</p>
+ *
+ * <p>After a drain if the internal queue has shrunk under {@link #lowWaterMark}, the queue is considered as {@code writable}.
+ * {@link #add}/{@link #drain} methods return {@link #QUEUE_WRITABLE_MASK} to signal producers should start emitting. Note
+ * the consumer thread handle this signal and should transmit it to the producers.</p>
+ *
+ * <p>When {@link #QUEUE_WRITABLE_MASK} is signalled, the number of {@link #QUEUE_UNWRITABLE_MASK} signals emitted is encoded
+ * in the flags. This number allows the producer flow controller to correctly account the producer writability:
+ * {@link #QUEUE_WRITABLE_MASK}/{@link #QUEUE_UNWRITABLE_MASK} signals are observed by different threads, therefore a good
+ * practice to compute the queue writability is to increment/decrement an atomic counter for each signal received.</p>
+ */
+public class OutboundWriteQueue<E> {
+
+  /**
+   * Returns the number of times {@link #QUEUE_UNWRITABLE_MASK} signals encoded in {@code value}
+   *
+   * @param value the value
+   * @return then number of unwritable signals
+   */
+  public static int numberOfUnwritableSignals(int value) {
+    return (value & ~0XF) >> 4;
+  }
+
+  /**
+   * When the masked bit is set, the queue became unwritable, this triggers only when the queue transitions
+   * from the <i>writable</i>> state to the <i>unwritable</i>> state.
+   */
+  public static int QUEUE_UNWRITABLE_MASK = 0x01;
```

Comment:
All the masks should be made static final

Reply:
good point, forgot it :-)